### ISSUE TITLE
One new function: last_over_time

### ISSUE THREAD

**Issue by alburthoffman** [prometheus/prometheus#7179]
## Proposal
**Use case. Why is this important?**

function last_over_time provides a solution to adjust the 5 minutes fall back behavior in prometheus. usage is like
```
last_over_time(metric{xxx}[10m])
```

In the document, there is
> If no sample is found (by default) 5 minutes before a sampling timestamp, no value is returned for that time series at this point in time. 

this is as same as below query:
```
last_over_time(metric{xxx}[5m])
```

use cases:

1. Graphs. 5m fallbacks always cause confusing, and needs double checks and calculation to confirm. This is a headache when triaging one issue, and easily make mistakes. 

1. Alerting. we simply hate false alerts, especially it's critical alerts. the 5m fallbacks will generate false alerts. Sometimes we want longer time range, like 10m downtime, then we can use 10m to tolerance instead of 5m. 

1. Calculation. When doing detection, precision is important. In order to get rid of 5m fallbacks, extra code is needed to check that. this function will simplify the queries.

Make a proposal here to see if it can help. PR can come later if we can make an agreement.

**Comment by brian-brazil** [comment]
This has already previously been discussed in #6530 and #1139, and this is not something we plan on adding.

**Comment by alburthoffman** [comment]
Thx. checked the two threads, and honestly I could not agree on the conclusion.
Added my comments in #6530 for this.

**Comment by brian-brazil** [comment]
Commenting on old closed issues is unlikely to change anything. If you wish for this to be considered you need to provide concrete use cases.

**Comment by alburthoffman** [comment]
One case is that we are doing abnormal detection by querying metrics from prometheus.  Since prometheus has 5 minutes fall back, then we have to write code to exclude those data points explicitly. 

Without doing this we will get false alerts and mislead people. few false alerts are ok, but when you have more than 10k rules running in ur system, you will do best to reduce the false alerts.

The other common case is to triage issue when querying metrics in Grafana. Most of time 5m fall back is OK, not a big deal. But when you want to align the time to confirm ur theory, it will mislead you.

@brian-brazil @roidelapluie

**Comment by brian-brazil** [comment]
> Since prometheus has 5 minutes fall back, then we have to write code to exclude those data points explicitly.

That's not how PromQL works, nor is that a concrete use case.

Can you explain exactly the queries you're using and what you're trying to do?

**Comment by alburthoffman** [comment]
we have queries like below. I simply use xxx to avoid exposing internal informations:

```
sum(last_over_time(bu_abnormal_detection{xxxx, type="severity"}[1m])) > 0 AND sum(last_over_time(bu_abnormal_detection{xxxx, type="raw"}[1m])) < sum(last_over_time(bu_abnormal_detection{xxxx, type="lower-bound"}[1m]))
```

bu_abnormal_detection is the metric calculated by other recording rules. 

In the case we have it's that the rules might be mis evaluated because of deployment or system issue.
If no last_over_time, this will generate false alerts and sent to SRE guys, they are not going to be happy with that. So we implement one function last_over_time and hack prometheus to use it.

**Comment by alburthoffman** [comment]
> That's not how PromQL works
Don't quite agreed on this. PromQL should return the metrics as they are. if no metrics, don't fall back.

**Comment by brian-brazil** [comment]
That's not enough to go on I'm afraid.

What happens if you do `sum((bu_abnormal_detection{xxxx, type="severity"}[1m])) > 0 AND sum((bu_abnormal_detection{xxxx, type="raw"}[1m])) < sum((bu_abnormal_detection{xxxx, type="lower-bound"}[1m]))`?

What's the actual math going on behind the covers here? Why might recent values be missing or stale?

**Comment by brian-brazil** [comment]
> PromQL should return the metrics as they are.

Prometheus is a monitoring system, not a generic TSDB. If you want raw data to do heavier statistical analysis then you should use something like R one the raw data from query_range.

> if no metrics, don't fall back.

That's how it's meant to work.

The only reason the 5m should be applying would be if you've a interval over 2m (which is not recommended), or you're exposing timestamps (which is not generally recommended).

**Comment by alburthoffman** [comment]
this kind of metrics are generated by one detection service
`bu_abnormal_detection{xxxx, type="lower-bound"}[1m]`

The service will return upper and lower bound by using some stats/ml models.

The thing is the models need historical data to train. And this is done by another recording rule because we want the aggregation results instead of simple raw metrics. 

the problem is sometimes the recording rule might be mis evaluated. and no lower and upper bound at that time, then it will compare with previous values, which will cause false alerts.

**Comment by brian-brazil** [comment]
> the problem is sometimes the recording rule might be mis evaluated. and no lower and upper bound at that time, then it will compare with previous values, which will cause false alerts.

Okay, this sounds like the root issue. If you can share the rules I can probably help with that.

**Comment by alburthoffman** [comment]
Since we have lots of rules running, we run prometheus as rule executor and query metrics from thanos cluster. It's a storage based system, not streaming base.

Rule set are changing from time to time. they might not run at one node all the time. If one node is down or unhealthy, then its rule set will be redeploy to other nodes.

Also metrics sending to Thanos might be delay as well.

**Comment by alburthoffman** [comment]
mis evaluation is few, but it will happen sometime.

In order to provide high SLA for business related alerts, we tried to find ways to reduce the false alerts

**Comment by brian-brazil** [comment]
Without you sharing your rules, I'm afraid I can't help you. Prometheus is also explicitly not a clustered query execution/storage engine. https://www.robustperception.io/why-can-countx-5-not-return-0 is what I suspect you're running into.

**Comment by roidelapluie** [comment]
We have looked into this issue in our bug scrub and decided to close this issue. We are willing to help you with your usecase if you can provide more data, as requested above. There might be other cases where last_over_time might be needed, but it does not look like this is one of them.

**Comment by roidelapluie** [comment]
This has been implemented in #8457 and will be available in Prometheus 2.26.
